### DISCUSSION TITLE
Delete related files on deletion of collection item

### DISCUSSION THREAD

**Discussion by domdus** [directus/directus#6283] [Feature Requests]
Hi,

my editors asked me about a "delete all related files on deletion of the collection item" feature.

So basically saying: If you don't need the meta - you don't need the files.

I guess that might be complicated if files in the scope for deletion are again linked to a totally different item. (where the files are needed)
Beside that such a feature (=> [x] Delete files checkbox in the dialog)  would help a lot with keeping the file lib clean from dead files.

**Comment by benhaynes** [comment]
I believe we already have this with the cascade delete on the foreign key constraints... but I deffer to @rijkvanzanten!

**Reply by rijkvanzanten** [reply]
Yes, but that only automatically deletes the record in the database, that will not delete the file on disk currently

**Comment by camopy** [comment]
Do you know if it is already possible to do?

**Reply by rijkvanzanten** [reply]
No, this is an open feature request 👍🏻

**Reply by camopy** [reply]
Cool! Thanks! Do you have the link?

**Reply by rijkvanzanten** [reply]
This is it!

**Reply by shaunchander** [reply]
I feel like this can be achieved with a Directus Flow **but** at the same time this should be natively supported (right now it's a one-way cascade but in the opposite direction: if you delete the file then you can delete the whole item).

**Reply by areejnaqvi** [reply]
@shaunchander hi, could you possibly explain how to do this using a directus flow? ive been trying to figure that out for a while now but i cant seem to get it work. thanks!

**Reply by domdus** [reply]
@areejnaqvi: Here is a guide on how to delete files when deleting of a collection item (that holds the files).
**The final flow:**
<img width="883" alt="image" src="https://github.com/directus/directus/assets/67951828/7cb391c0-7f6e-43b6-8f28-70b843d7172f">

---
1. Create a flow that triggers on delete of collection items. It needs to be of type Filter (Blocking) in order to get the file id from the collection item BEFORE it gets deleted.
<img width="527" alt="image" src="https://github.com/directus/directus/assets/67951828/635844c1-ef7a-4a2d-9ad0-5d228ffa4eea">
Don't forget to select the scope collection...

---
When the flow triggers you will get the item id in the payload of the flow (see logs). Also notice the scope collection named `test` in this example.
<img width="539" alt="image" src="https://github.com/directus/directus/assets/67951828/61127d7e-8471-42ef-b52b-8a0e4b309479">

2. Add a Read Data operation to get the item data (file id)
-  Select collection "test" (<-- or you could make it dyamic using value of `{{$trigger.collection}}` to cover mutliple collections in one flow (test, abc, xyz))
- For the query you apply a filter for the item id and use a dynamic value from the triggers payload:
`{{$trigger.payload}}` (payload is an array)
<img width="422" alt="image" src="https://github.com/directus/directus/assets/67951828/e4b65281-2cad-452c-8cc1-dc554d1ef4a9">

---
Now when the flow triggers you should find data in the Read Data operation payload - as well as the `file`-field. (named "file" here)
<img width="538" alt="image" src="https://github.com/directus/directus/assets/67951828/1b57f7ea-6a55-4122-8cd2-8ad3bdba7738">

3. In order to get it fully working also for batch item deletions we need a "Run Script"-operation as next. It will prepare a webhook payload. (see 4.)

In Run-Script we iterate the items (1-n) we get from the last operation payload (array from Read Data) and collect the file ids of the `file`-field:
```
module.exports = async function(data) {
    if (data?.$last.length > 0) { // check if payload from $last operation is not empty
        const ids = [];
        data?.$last.map(item => {
             // add file-field (asset) id to array of ids
             ids.push(item?.file); 

             // it can be another additional file field in your data model
             // e.g. ids.push(item?.my_image_field);
             //        ids.push(item?.my_whatever_file_field); 
        })
        return ids; // This will be the ready-to-go-array of file ids for the DELETE request body
    }
    return {};
}
```
<img width="536" alt="image" src="https://github.com/directus/directus/assets/67951828/32b6265c-2ab2-427d-abb6-5cfea4f4e948">

This will return an array of file ids we want to delete.

4. There is no "Delete File"-operation that I'm aware of. So to delete the files we need to call the directus /files api endpoint via a webhook DELETE request.
https://docs.directus.io/reference/files.html#delete-multiple-files

The /files api DELETE methods accepts an array of keys (file ids), which is exactly the payload we prepared in the $last operation (Run Script) - the ready-to-go-array.
Just paste {{$last}} in the request body:
<img width="421" alt="image" src="https://github.com/directus/directus/assets/67951828/baf15b9a-d864-4339-907d-23a049287dc3">
As well you need to have a token from a user that is allowed to delete files (directus_files) via the api.
- Create a user and give it a static token
- Assign a role that allows to delete in directus_files (or make it an admin?)
- Add header Authorization and the static token:

---
And that's it!
Now when for example 3 collection items get deleted:
1. (blocking) filter flow triggers (3 collection item ids in payload)
2. Read Data queries the 3 items one last time (get all fields) 
3. Run Script processes the payload (extract/collect file ids) and provides the result array for the webhook
4. Webhook deletes the corresponding files in a single api request

<img width="527" alt="image" src="https://github.com/directus/directus/assets/67951828/9bbc13c3-97f2-47e8-9670-9697e7437271">

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
